src/runtime_src/tools/xclbinutil/SectionVenderMetadata.cxx

```diff
@@ -0,0 +1,354 @@
+  return  (m_pBuffer != nullptr);
+}
+
+// -------------------------------------------------------------------------
+
+bool
+SectionVenderMetadata::supportsSubSection(const std::string& _sSubSectionName) const
+{
+  // There is only one-subsection that is supported.  By default it is not named.
+  return _sSubSectionName.empty();
+}
+
+
+// -------------------------------------------------------------------------
+void
+SectionVenderMetadata::copyBufferUpdateMetadata(const char* _pOrigDataSection,
+                                                unsigned int _origSectionSize,
+                                                std::istream& _istream,
+                                                std::ostringstream& _buffer) const
+{
+  XUtil::TRACE("SectionVenderMetadata::CopyBufferUpdateMetadata");
+
+  // Do we have enough room to overlay the header structure
+  if (_origSectionSize < sizeof(vender_metadata)) {
+    const auto& errMsg = boost::str(boost::format(
+                                        "ERROR: Segment size (%d) is smaller than the size of the vender_metadata structure (%d)")
+                                    % _origSectionSize % sizeof(vender_metadata));
+    throw std::runtime_error(errMsg);
+  }
+
+  // Prepare our destination header buffer
+  vender_metadata venderMetadataHdr = { 0 };   // Header buffer
+  std::ostringstream stringBlock;              // String block (stored immediately after the header)
+
+  const vender_metadata* pHdr = reinterpret_cast<const vender_metadata*>(_pOrigDataSection);
+
+  XUtil::TRACE_BUF("vender_metadata-original", reinterpret_cast<const char*>(pHdr), sizeof(vender_metadata));
+  XUtil::TRACE(boost::str(boost::format(
+                              "Original: \n"
+                              "  mpo_name (0x%lx): '%s'\n"
+                              "  m_image_offset: 0x%lx, m_image_size: 0x%lx\n")
+                          % pHdr->mpo_name % (reinterpret_cast<const char*>(pHdr) + pHdr->mpo_name)
+                          % pHdr->m_image_offset % pHdr->m_image_size));
+
+  // Get the JSON metadata
+  _istream.seekg(0, _istream.end);             // Go to the beginning
+  std::streampos fileSize = _istream.tellg();  // Go to the end
+
+  // Copy the buffer into memory
+  std::unique_ptr<unsigned char> memBuffer(new unsigned char[fileSize]);
+  _istream.clear();                                // Clear any previous errors
+  _istream.seekg(0);                               // Go to the beginning
+  _istream.read((char*)memBuffer.get(), fileSize); // Read in the buffer into memory
+
+  XUtil::TRACE_BUF("Buffer", (char*)memBuffer.get(), fileSize);
+
+  // Convert JSON memory image into a boost property tree
+  std::stringstream ss;
+  ss.write((char*)memBuffer.get(), fileSize);
+
+  boost::property_tree::ptree pt;
+  boost::property_tree::read_json(ss, pt);
+
+  // Extract and update the data
+  boost::property_tree::ptree& ptSK = pt.get_child("vender_metadata");
```

Comment:
is there a scenario when this tree is not present? If so, maybe we should have the pt_empty check